nx/lib/nx/defn/compiler.ex

```diff
@@ -33,6 +33,25 @@ defmodule Nx.Defn.Compiler do
               opts :: keyword
             ) :: [Nx.Container.t()]
 
+  @doc """
+  Callback for compilation.
+
+  It receives the function `vars`, the function `fun` which builds
+  a defn expression, and the compiler options. It must call `fun`
+  with the `vars` as arguments.
+
+  It returns a function that receives a list of arguments and
+  returns a list of results.
+
+  The callback uses double underscores so it can be defined
+  at root modules without affecting the module's main API.
+  """
+  @callback __compile__(
+              vars :: [Nx.Container.t()],
+              fun :: ([Nx.Container.t()] -> Nx.Container.t()),
+              opts :: keyword
+            ) :: ([[Nx.t()]] -> [Nx.Container.t()])
```

Comment:
Can't the input also be `Nx.Container.t()`?

Reply:
No, we flatten all containers before.

Reply:
Makes sense, thanks!